opentech/apply/utils/blocks.py

```diff
@@ -0,0 +1,163 @@
+        'style_formats': [
+            {'title': 'Headers', 'items': [
+                {'title': 'Header 1', 'format': 'h1'},
+                {'title': 'Header 2', 'format': 'h2'},
+                {'title': 'Header 3', 'format': 'h3'},
+            ]},
+            {'title': 'Inline', 'items': [
+                {'title': 'Bold', 'icon': 'bold', 'format': 'bold'},
+                {'title': 'Italic', 'icon': 'italic', 'format': 'italic'},
+                {'title': 'Underline', 'icon': 'underline', 'format': 'underline'},
+            ]},
+        ],
+    })
+
+    class Meta:
+        label = _('Rich text field')
+        icon = 'form'
+
+    def get_searchable_content(self, value, data):
+        return bleach.clean(data, tags=[], strip=True)
+
+
+class CustomFormFieldsBlock(StreamBlock):
+    rich_text = RichTextFieldBlock(group=_('Fields'))
+    required_blocks = None
+    required_block_names = None
+
+    def __init__(self, *args, **kwargs):
+        child_blocks = [(block.name, block(group=_('Required'))) for block in self.required_blocks]
+        self.required_block_names = [block.name for block in self.required_blocks]
+
+        super().__init__(child_blocks, *args, **kwargs)
+
+    def clean(self, value):
+        try:
+            value = super().clean(value)
+        except ValidationError as e:
+            error_dict = e.params
+        else:
+            error_dict = dict()
+
+        block_types = [block.block_type for block in value]
+        missing = set(self.required_block_names) - set(block_types)
+
+        duplicates = [
+            name for name in find_duplicates(block_types)
+            if name in self.required_block_names
+        ]
+
+        all_errors = list()
+        if missing:
+            all_errors.append(
+                'You are missing the following required fields: {}'.format(', '.join(prettify_names(missing)))
+            )
+
+        if duplicates:
+            all_errors.append(
+                'You have duplicates of the following required fields: {}'.format(', '.join(prettify_names(duplicates)))
+            )
+            for i, block_name in enumerate(block_types):
+                if block_name in duplicates:
+                    self.add_error_to_child(error_dict, i, 'info', 'Duplicate field')
+
+        if all_errors or error_dict:
+            error_dict['__all__'] = all_errors
+            raise ValidationError('Error', params=error_dict)
+
+        return value
+
+    def add_error_to_child(self, errors, child_number, field, message):
+        new_error = ErrorList([message])
+        try:
+            errors[child_number].data[0].params[field] = new_error
+        except KeyError:
+            errors[child_number] = ErrorList(
+                [ValidationError('Error', params={field: new_error})]
+            )
+
+    def to_python(self, value):
+        # If the data type is missing, fallback to a CharField
```

Comment:
Can we add a docstring to this to say that this allows historic data to still be accessible even if a custom field type is removed from the code in the future.

(I should have done that originally)